### PR TITLE
MGMT-14276: (RecoveryISO) download ocp release mirror

### PR DESCRIPTION
**Pull request by nmagnezi**
@danielerez

### CODE REVIEW

File: pkg/asset/recovery/recovery_iso.go [hunk 1]
```diff
@@ -86,3 +96,47 @@ func (a *RecoveryISO) Generate(dependencies asset.Parents) error {
 func (a *RecoveryISO) Name() string {
 	return "Appliance Recovery ISO"
 }
+
+func (a *RecoveryISO) ocpMirror(AssetsDir string, dependencies asset.Parents) error {
```
Comment (danielerez) [inline pkg/asset/recovery/recovery_iso.go:100]: This func can be added as an API in release.go (add to interface, etc)

File: pkg/asset/recovery/recovery_iso.go [hunk 2]
```diff
@@ -86,3 +96,47 @@ func (a *RecoveryISO) Generate(dependencies asset.Parents) error {
...
+	applianceConfig := &config.ApplianceConfig{}
+	dependencies.Get(&config.EnvConfig{}, applianceConfig)
+	exec := &executer.CommonExecuter{}
+
+	ps, err := exec.TempFile(AssetsDir, "pull-secret")
```
Comment (danielerez) [inline pkg/asset/recovery/recovery_iso.go:105]: Forgot we already have this logic in execute :) https://github.com/danielerez/openshift-appliance/blob/59d005ef5aa192c855396709968df54051a8ae83/pkg/release/release.go#L118-L119

File: pkg/asset/recovery/recovery_iso.go [hunk 3]
```diff
@@ -44,18 +50,22 @@ func (a *RecoveryISO) Generate(dependencies asset.Parents) error {
 	baseISO := &BaseISO{}
 	dependencies.Get(envConfig, baseISO)
 	coreosIsoPath := filepath.Join(envConfig.CacheDir, coreosIsoFileName)
-	recoveryIsoDirPath := filepath.Join(envConfig.CacheDir, recoveryIsoDirName)
+	recoveryIsoPath := filepath.Join(envConfig.CacheDir, recoveryIsoDirName)
 
-	if err := os.MkdirAll(recoveryIsoDirPath, os.ModePerm); err != nil {
-		logrus.Errorf("Failed to create dir: %s", recoveryIsoDirPath)
+	if err := os.MkdirAll(recoveryIsoPath, os.ModePerm); err != nil {
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply pkg/asset/recovery/recovery_iso.go:64]: Done

File: pkg/asset/recovery/recovery_iso.go (hunk 1 above)
Comment (nmagnezi) [reply pkg/asset/recovery/recovery_iso.go]: Done

File: pkg/asset/recovery/recovery_iso.go [hunk 4]
```diff
@@ -1,20 +1,25 @@
...
 )
 
 const (
-	coreosIsoFileName   = "coreos-x86_64.iso"
-	recoveryIsoFileName = "recovery.iso"
-	recoveryIsoDirName  = "recovery_iso"
+	coreosIsoFileName     = "coreos-x86_64.iso"
+	recoveryIsoFileName   = "recovery.iso"
+	recoveryIsoDirName    = "recovery_iso"
+	ocAdmMirrorCmd        = "oc adm release mirror --from=%s --to-dir %s --registry-config=%s"
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply pkg/asset/recovery/recovery_iso.go]: Done

File: pkg/asset/appliance/base_diskimage.go [hunk 5]
```diff
@@ -33,8 +33,8 @@ func (a *BaseDiskImage) Generate(dependencies asset.Parents) error {
 	applianceConfig := &config.ApplianceConfig{}
 	dependencies.Get(envConfig, applianceConfig)
 
-	c := coreos.NewCoreOS(envConfig.CacheDir)
-	cpuArch, err := a.getCpuArch(applianceConfig, envConfig.CacheDir)
+	c := coreos.NewCoreOS(envConfig.CacheDir, envConfig.AssetsDir, envConfig.TempDir)
```
Comment (danielerez) [inline pkg/asset/appliance/base_diskimage.go:36]: you can just pass the entire envConfig here, will look a bit cleaner:)

File: pkg/asset/recovery/recovery_iso.go [hunk 6]
```diff
@@ -31,31 +34,52 @@ var _ asset.Asset = (*RecoveryISO)(nil)
...
 	coreosIsoPath := filepath.Join(envConfig.CacheDir, coreosIsoFileName)
-	recoveryIsoDirPath := filepath.Join(envConfig.TempDir, recoveryIsoDirName)
+	recoveryIsoPath := filepath.Join(envConfig.TempDir, recoveryIsoDirName)
+	c := coreos.NewCoreOS(envConfig.CacheDir, envConfig.AssetsDir, envConfig.TempDir)
+
+	// Search for disk image in cache dir
+	if fileName := c.FindInCache(recoveryIsoFileName); fileName != "" {
+		logrus.Info("Reusing recovery ISO from cache...")
+		a.File = &asset.File{Filename: fileName}
+		return nil
```
Comment (danielerez) [inline pkg/asset/recovery/recovery_iso.go:58]: We actually don't want to return here, as the next step is to embed the ignition. I.e. as embedding the ignition the a quick operation, we can always perform without rebuilding the recovery iso again.

File: pkg/release/release.go [hunk 7]
```diff
@@ -32,27 +32,36 @@ const (
...
+	releaseImage string
+	pullSecret   string
+	cacheDir     string
+	assetsDir    string
+	tempDir      string
 }
 
 // NewRelease is used to set up the executor to run oc commands
-func NewRelease(executer executer.Executer, releaseImage, pullSecret, cacheDir string) Release {
+func NewRelease(executer executer.Executer, releaseImage, pullSecret, cacheDir, assetsDir, tempDir string) Release {
```
Comment (danielerez) [inline pkg/release/release.go:48]: It's getting a bit too long:) Maybe just pass the entire envConfig here instead.

File: pkg/release/release.go [hunk 8]
```diff
@@ -32,27 +32,36 @@ const (
 type Release interface {
 	ExtractFile(image string, filename string) (string, error)
 	GetReleaseArchitecture() (string, error)
+	OcpMirror(ocpReleaseImage string) error
```
Comment (danielerez) [inline pkg/release/release.go:35]: Just for convention, let's call the func 'MirrorRelease'

Comment (nmagnezi) [reply pkg/release/release.go]: Done

File: pkg/release/release.go (hunk 7 above)
Comment (nmagnezi) [reply pkg/release/release.go]: Done

File: pkg/asset/recovery/recovery_iso.go (hunk 6 above)
Comment (nmagnezi) [reply pkg/asset/recovery/recovery_iso.go]: Done

**Pull request merged**